File: src/go/4_BehavioralPatterns/11_Visitor/intrusive.go
package visitor

import (
  "fmt"
  "strings"
)

type Expression interface {
  Print(sb *strings.Builder)
}

type DoubleExpression struct {
  value float64
}

func (d *DoubleExpression) Print(sb *strings.Builder) {
  sb.WriteString(fmt.Sprintf("%g", d.value))
}

type AdditionExpression struct {
  left, right Expression
}

func (a *AdditionExpression) Print(sb *strings.Builder) {
  sb.WriteString("(")
  a.left.Print(sb)
  sb.WriteString("+")
  a.right.Print(sb)
  sb.WriteString(")")
}

func main() {
  // 1+(2+3)
  e := AdditionExpression{
    &DoubleExpression{1},
    &AdditionExpression{
      left:  &DoubleExpression{2},
      right: &DoubleExpression{3},
    },
  }
  sb := strings.Builder{}
  e.Print(&sb)
  fmt.Println(sb.String())
}

File: src/js/4_BehavioralPatterns/11_Visitor/intrusive.js
class NumberExpression
{
  constructor(value)
  {
    this.value = value;
  }

  print(buffer)
  {
    buffer.push(this.value.toString());
  }
}

class AdditionExpression
{
  constructor(left, right)
  {
    this.left = left;
    this.right = right;
  }

  print(buffer)
  {
    buffer.push('(');
    this.left.print(buffer);
    buffer.push('+');
    this.right.print(buffer);
    buffer.push(')');
  }
}

// 1 + (2+3)
let e = new AdditionExpression(
  new NumberExpression(1),
  new AdditionExpression(
    new NumberExpression(2),
    new NumberExpression(3)
  )
);
let buffer = [];
e.print(buffer);
console.log(buffer.join(''));